File: tests/TestBindings/TestRace.cpp
#include <catch.hpp>
#include <sol.hpp>
#include <BWAPI_Lua.h>
#include "BindingTestHelpers.h"

using namespace BWAPI_Lua;

TEST_CASE("race bindings", "[race]")
{
	sol::state lua;
	lua.open_libraries(sol::lib::base);
	bindRace(lua.create_named_table("BWAPI"));
	lua.script_file(TestBindings::getLuaTestPath(__FILE__));

	SECTION("constructors")
	{
		sol::function test = lua["constructors"];
		REQUIRE_NOTHROW(test());
	}

	SECTION("equality")
	{
		sol::function test = lua["equality"];
		REQUIRE_NOTHROW(test());
	}

	SECTION("tostring")
	{
		sol::function test = lua["toStringTest"];
		REQUIRE_NOTHROW(test());
	}

	SECTION("race values")
	{
		bindUnitType(lua["BWAPI"]);

		SECTION("Zerg")
		{
			sol::function test = lua["valuesZerg"];
			REQUIRE_NOTHROW(test());
		}

		SECTION("Unknown")
		{
			sol::function test = lua["valuesUnknown"];
			REQUIRE_NOTHROW(test());
		}
	}
}

File: tests/TestBindings/Lua/TestRace.lua
function constructors()
  race = BWAPI.Race(BWAPI.Races.Enum.Zerg)
end

function equality()
  race = BWAPI.Race(BWAPI.Races.Enum.Zerg)
  unknown = BWAPI.Race(BWAPI.Races.Enum.Unknown)

  assert(race ~= unknown)
  assert(race == BWAPI.Races.Zerg)
  assert(unknown == BWAPI.Races.Unknown)
end

function toStringTest()
  race = BWAPI.Race(BWAPI.Races.Enum.Zerg)
  unknown = BWAPI.Race(BWAPI.Races.Enum.Unknown)

  assert(tostring(race) == 'Zerg')
  assert(tostring(unknown) == 'Unknown')
end

function valuesZerg()
  race = BWAPI.Race(BWAPI.Races.Enum.Zerg)

  assert(race:getWorker() == BWAPI.UnitTypes.Zerg_Drone)
  assert(race:getCenter() == BWAPI.UnitTypes.Zerg_Hatchery)
  assert(race:getRefinery() == BWAPI.UnitTypes.Zerg_Extractor)
  assert(race:getTransport() == BWAPI.UnitTypes.Zerg_Overlord)
  assert(race:getSupplyProvider() == BWAPI.UnitTypes.Zerg_Overlord)
end

function valuesUnknown()
  race = BWAPI.Races.Unknown

  assert(race:getWorker() == BWAPI.UnitTypes.Unknown)
  assert(race:getCenter() == BWAPI.UnitTypes.Unknown)
  assert(race:getRefinery() == BWAPI.UnitTypes.Unknown)
  assert(race:getTransport() == BWAPI.UnitTypes.Unknown)
  assert(race:getSupplyProvider() == BWAPI.UnitTypes.Unknown)
end
